### PR TITLE
OAuth Server

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Initial implementation of an OAuth server in GLPI built on top of the new High-Level API using `league/oauth2-server`.
Current grants supported:
- Password
- Client Credentials

Currently, access tokens not associated with a user will not be considered valid by the HL API endpoints (client credential grant for example) but they are still considered valid by the OAuth server itself. Since the API requires a user, it uses the user's permissions still and doesn't have any scope support currently.

### CODE REVIEW

**Comment by cconard96** [comment]
How to use in Swagger UI with the new API (after updating database schema for this PR):
1. Make sure your profile has the required rights for OAuth Clients in the Setup tab
2. Go to Setup > OAuth Clients and click the "Add" button to create a new client
3. Optionally fill out the name and comments field, and select at least the Password grant type in the Grants field, then click "Add"
4. Open the newly created OAuth Client
5. In the Swagger UI, click the "Authorize" button in the top right of the Swagger UI page `/api.php/doc`
6. For the OAuth authentication section, enter your GLPI username, GLPI password, the client ID and client secret from the OAuth Client form, and change the "Client credentials location" to request body.
7. Click the authorize button

You will then be automatically authenticated for any endpoints you try from the Swagger UI page.

**Comment by orthagh** [comment]
Seems early for review, but here is some things i noticed:

- Table name seems inconsistent, install/update create a table named `glpi_oauth_clients`, but the current code ask for `glpi_oauthclients`. I renamed it to continue to continue my tests.
- Client secret should be hashed in database. On UX perspective, the secret should be given to the user at creation (and warn we can't display it afterward). This may requires also a mechanism to regenerate it (a button next to it ?).
- Client name field could benefit to a copy control like password field
- Grants field may have a default value (all ?). I missed it.
- About field organization in UI, i guess we may have client id and secret one above the other (maybe set the grant field before them)

Suggestions:
- add an enabled field (on by default)
- An expiration date field could be a plus (empty by default)
- We may add an optional logging system (in a tab) to track auth usage (maybe organized by sessions). We already did this for apiclient in legacy API.
- About that, the new oauth client is somehow a duplicated feature with apiclient. Maybe we can merge them somehow and have a flag/selection to indicate if we want legacy authorization or new oauth one.
- we may want a scope system. For the moment only new api uses the tokens, but we may extend this authentication to inventory api and legacy api

**Comment by cconard96** [comment]
>     * Client secret should be hashed in database. On UX perspective, the secret should be given to the user at creation (and warn we can't display it afterward). This may requires also a mechanism to regenerate it (a button next to it ?).

It was supposed to be encrypted on add/update and decrypted after getting the client from the DB. Indeed, showing it only once would be preferred.

>     * We may add an optional logging system (in a tab) to track auth usage (maybe organized by sessions). We already did this for apiclient in legacy API.

Not sure about "organized by sessions". REST is supposed to be stateless and currently when you authenticate with OAuth to the API, a new temporary session is started instead of one meant to be persisted and reused. If we want persistent sessions with OAuth, the session ID would need to be saved along with the access token the first time it is used. I added support to specify `Glpi-Entity` and `Glpi-Profile` request headers to set the entity and profile for each request instead of using the server session if we stick with a more stateless design.

>     * we may want a scope system. For the moment only new api uses the tokens, but we may extend this authentication to inventory api and legacy api

The inventory submission is currently going to a `front` file. Shouldn't it be moved to the new API anyways? Scopes should already be supported by the DB schema, OAuth client UI, and OAuth server implementation. It is just a matter of defining the scopes in the code.

**Comment by orthagh** [comment]
Another thing we didn't discussed, the secondary goal for the feature login delegation from other applications (like keycloak can do). 
Is anything already available for that ?

**Comment by cconard96** [comment]
I don't know KeyCloak. Are you referring to the authorization code grant type?

https://oauth2.thephpleague.com/authorization-server/auth-code-grant/

If so, that grant isn't enabled and I need to do more research to set it up properly.

**Comment by orthagh** [comment]
> I don't know KeyCloak

you can install on docker. It's pretty straighforward.


>  authorization code 

yes exactly

**Comment by cconard96** [comment]
> > authorization code
> 
> yes exactly

It is added now and seems to work OK within the Swagger UI page at least after many, many tries.

**Comment by trasher** [comment]
Are we sure we should make `ext-sodium` required? Mot of other requirements are kind of "standard", this one is probably not. 
Also, the polyfill is provided, the only issue could be a perf issue; is there really a perf issue without the extension on this pull request?

**Comment by cconard96** [comment]
> Are we sure we should make `ext-sodium` required? Mot of other requirements are kind of "standard", this one is probably not. Also, the polyfill is provided, the only issue could be a perf issue; is there really a perf issue without the extension on this pull request?

I was mainly trying to figure out why the CI actions were failing on the build dependencies step.

**Comment by trasher** [comment]
> Are we sure we should make `ext-sodium` required?

Plus it will be required to instal the extension even if user does not plan to use oauth server

**Comment by cedric-anne** [comment]
> > Are we sure we should make `ext-sodium` required? Mot of other requirements are kind of "standard", this one is probably not. Also, the polyfill is provided, the only issue could be a perf issue; is there really a perf issue without the extension on this pull request?
> 
> I was mainly trying to figure out why the CI actions were failing on the build dependencies step.

Best solution is ta make this extension marked as provided by the polyfill, see https://github.com/paragonie/sodium_compat/pull/156

I also recently asked myself how we should handle extensions in CI env. If we would have all extensions installed, we would not have see this issue. We could have polyfilled extensions removed, to detect probable issues, but it may have a negative impact on tests. A solution could be to have all extensions installed on default CI tests, and use polyfills on additonnal tests that are made optionnaly (on merge in main branches, nightly checks, ...).

**Comment by trasher** [comment]
Testing presence of any extension sounds useless. Checks are done on upgrade and install, for most of the case this will be enough. 
And we should not install optional extensions on our CI if it's not strictly required (such as ldap ext for tests).

The problem here is a lack of proper configuration on 3rd party lib; this kind of issue is not seen often.

File: composer.json
Comment (cedric-anne) [inline composer.json:96]: I added this requirement to be able to run tests. If https://github.com/paragonie/sodium_compat/pull/156 is accepted, we will have to remove it.

File: install/install.php
Comment (cedric-anne) [inline install/install.php:448]: Calls to this method could be placed after calls to `RulesManager::initializeRules();`, after the install/update process.

File: composer.json
Comment (cedric-anne) [reply composer.json:96]: Solution was refused by `paragonie/sodium_compat` maintainers because their polyfill is not covering 100% of the extension.
We have to keep this workaround.

File: src/System/RequirementsManager.php
Comment (cedric-anne) [inline src/System/RequirementsManager.php:117]: Do you think it could be possible to make this requirement optional? Maybe we could "disable" the OAuth server / API when extension is not loaded, and add a message like "To use XXX, you must first enable the opensssl PHP extension." in related pages.

File: inc/based_config.php
Comment (cedric-anne) [inline inc/based_config.php:134]: Not sure these kind of constants are really usefull.

File: install/migrations/update_10.0.x_to_10.1.0/oauth.php
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/oauth.php:107]: If keys are stored on filesystem, what would happen if someone forgot to copy them when reinstalling GLPI on another server?

Anyways, storing them in DB does not seems secure, so we should probably keep them here.

Maybe code should be moved in `Update::doUpdates()`, to ensure that missing keys would be regenerated after migrations (like we do for encryption key).

File: install/mysql/glpi-empty.sql
Comment (cedric-anne) [inline install/mysql/glpi-empty.sql:9512]:
```suggestion
```

File: src/Api/HL/Controller/CoreController.php
Comment (cedric-anne) [inline src/Api/HL/Controller/CoreController.php:453]: Redirection should probably be done using an absolute URL. Indeed, GLPI can be located on a sub directory.

File: src/OAuth/Server.php
Comment (cedric-anne) [inline src/OAuth/Server.php:177]: Code must be as defensive as possible here. Many methods can return `false` if something went wrong, and this should probably trigger an exception or make this method return false.

File: src/OAuthClient.php
Comment (cedric-anne) [inline src/OAuthClient.php:132]: Is it a good idea to make it decryptable? Most OAuth provider are only displaying this code once, and I guess they are hashing them.

If we want to make it decryptable, it should be added to `GLPIKey::$fields`, to make sure it is updated when encryption key is regenerated.

File: src/System/RequirementsManager.php
Comment (cconard96) [reply src/System/RequirementsManager.php:117]: Is it not included with default distributions of PHP?
At least for me, there wasn't a separate extension to install or enable.

Comment (cedric-anne) [reply src/System/RequirementsManager.php:117]: I do not really know, this point should be checked. Anyway, it can be done in another PR, if we figure out that this extension is not available on default PHP distributions for some OS.

File: src/OAuth/Server.php
Comment (cedric-anne) [inline src/OAuth/Server.php:178]:
```suggestion
                if ($pubkey !== false && file_put_contents($public_key_path, $pubkey['key']) === strlen($pubkey['key'])) {
```

File: src/OAuthClient.php
Comment (cedric-anne) [inline src/OAuthClient.php:149]:
```suggestion
```

File: src/OAuthClient.php
Comment (cedric-anne) [inline src/OAuthClient.php:114]:
```suggestion
        $input['secret'] = password_hash(self::getNewIDOrSecret(), PASSWORD_DEFAULT);
```

File: src/OAuth/ClientRepository.php
Comment (cedric-anne) [inline src/OAuth/ClientRepository.php:72]:
```suggestion
        if (password_verify($clientSecret, $client->fields['secret'])) {
```

**Review by cedric-anne — COMMENTED** [review]
Hashing `secret` would be more secure. It requires UI to be changed to be able to view/copy the `secret` only when it has just been created.

File: src/OAuthClient.php
Comment (cedric-anne) [reply src/OAuthClient.php:132]: We discussed about this in our weekly meeting and we concluded that we could keep the current logic based on decryptable secrets.

Only required change here is to register this field in `GLPIKey::$fields` to ensure value will be correctly re-encrypted when encryption key is changed.

**Comment by trasher** [comment]
I have no idea how to test that, I cannot review

**Comment by cedric-anne** [comment]
I rebased, squashed commits and upgraded `lcobucci/jwt` to 5.0 and `league/oauth2-server` to 8.5.

**Comment by cconard96** [comment]
> I have no idea how to test that, I cannot review

This is used by the new API. Using Insomnia or Postman should handle even the authorization code flow well, but the easiest method to test is probably from the Swagger documentation page (/api.php/doc).

You can create an OAuth client in Setup > OAuth Clients.

On the Swagger page:
1. Click the Authorize button (green outline in the top right)
2. Under "oauth (OAuth2, authorizationCode)", enter the OAuth client ID and Secret
3. Click Authorize to be redirected to the GLPI login/authorization prompt
4. After going through the authorization process, you should be able to use any of the endpoints in the Swagger UI
5. Refresh the page
6. Click the Authorize button (green outline in the top right)
7. Under "oauth (OAuth2, password)", enter the OAuth client ID, Secret, GLPI username, and GLPI password
8. Click Authorize
9. You should be able to use any of the endpoints in the Swagger UI

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]
Ok for me.

One last suggestion (could be done in a separate PR),  Grants field could be init with authorizationCode. I missed the field on my current review test, and i guess we can set the most secure grant by default

**Comment by cedric-anne** [comment]
> Ok for me.
> 
> One last suggestion (could be done in a separate PR), Grants field could be init with authorizationCode. I missed the field on my current review test, and i guess we can set the most secure grant by default

@cconard96 

I will merge this one once tests will be passed. It will have to be done in another PR.

**Pull request merged**
